plugin/pkg/scheduler/algorithm/priorities/interpod_affinity.go

```diff
@@ -0,0 +1,204 @@
+	nodeLister algorithm.NodeLister
+}
+
+func NewInterPodAffinityPriority(nodeLister algorithm.NodeLister) algorithm.PriorityFunction {
+	interPodAffinity := &InterPodAffinity{
+		nodeLister: nodeLister,
+	}
+	return interPodAffinity.CalculateInterPodAffinityPriority
+}
+
+// countPodsThatMatchPodAffinityTerm counts the number of given pods that match the podAffinityTerm.
+func countPodsThatMatchPodAffinityTerm(pod *api.Pod, podsOnNode []*api.Pod, node *api.Node, podAffinityTerm api.PodAffinityTerm) (int, error) {
+	labelSelector, err := unversioned.LabelSelectorAsSelector(podAffinityTerm.LabelSelector)
+	if err != nil {
+		return 0, err
+	}
+
+	matchedCount := 0
+	names := priorityutil.GetNamespacesFromPodAffinityTerm(pod, podAffinityTerm)
+	filteredPods := priorityutil.FilterPodsByNameSpaces(names, podsOnNode)
+	for _, filteredPod := range filteredPods {
+		if labelSelector.Matches(labels.Set(filteredPod.Labels)) {
+			if priorityutil.IfNodeHasTopologyKey(node, podAffinityTerm.TopologyKey) {
+				matchedCount++
+			}
+		}
+	}
+	return matchedCount, nil
+}
+
+type TopologyCounts map[string]int
+
+// if the topology key is empty, count weight for each node label value as index,
+// otherwise only count weight for node.Labels[topologyKey] as index.
+func (tc TopologyCounts) CountWeightByTopologykey(node *api.Node, weight int, topologyKey string) {
+	if len(topologyKey) == 0 && node.Labels != nil {
+		for _, index := range node.Labels {
+			tc[index] += weight
+		}
+	} else if len(topologyKey) > 0 && node.Labels != nil && len(node.Labels[topologyKey]) > 0 {
+		index := node.Labels[topologyKey]
+		tc[index] += weight
+	}
+	return
+}
+
+// CountWeightByPodMatchAffinityTerm counts the weight to topologyCounts for all the given pods that match the podAffinityTerm.
+func (tc TopologyCounts) CountWeightByPodMatchAffinityTerm(pod *api.Pod, podsOnNode []*api.Pod, weight int, podAffinityTerm api.PodAffinityTerm, node *api.Node) error {
+	if weight == 0 {
+		return nil
+	}
+	// get the pods which are there in that particular node
+	podsMatchedCount, err := countPodsThatMatchPodAffinityTerm(pod, podsOnNode, node, podAffinityTerm)
+	if err != nil {
+		return err
+	}
+	if podsMatchedCount > 0 {
+		tc.CountWeightByTopologykey(node, weight*podsMatchedCount, podAffinityTerm.TopologyKey)
+	}
+	return nil
+}
+
+// compute a sum by iterating through the elements of weightedPodAffinityTerm and adding
+// "weight" to the sum if the corresponding PodAffinityTerm is satisfied for
+// that node; the node(s) with the highest sum are the most preferred.
+// Symmetry need to be considered for preferredDuringSchedulingIgnoredDuringExecution from podAffinity & podAntiAffinity,
+// symmetry need to be considered for hard requirements from podAffinity
+func (s *InterPodAffinity) CalculateInterPodAffinityPriority(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, nodeLister algorithm.NodeLister) (schedulerapi.HostPriorityList, error) {
```

Comment:
This has a similar issue as the hard affinity and hard anti-affinity. What you want is something like this (obviously this is just pseudocode, and I've omitted some stuff but hopefully it will give you the idea)

```
func CountWeightByPodMatchAffinityTerm(pod *api.Pod, allPods []*api.Pod, weight int, podAffinityTerm api.PodAffinityTerm, node *api.Node) (int, error) {
      podsMatchedCount, err := countPodsThatMatchPodAffinityTerm(pod, allPods, node, podAffinityTerm)
      return weight * podsMatchedCount, nil
}

func (s *InterPodAffinity) CalculateInterPodAffinityPriority(pod *api.Pod, nodeNameToInfo map[string]*schedulercache.NodeInfo, nodeLister algorithm.NodeLister) (schedulerapi.HostPriorityList, error) {
   for _, node := range nodes.Items {
      total int  // total priority for this node
      // count weights for the weighted pod affinity
      if affinity.PodAffinity != nil {
         for _, weightedTerm := range affinity.PodAffinity.PreferredDuringSchedulingIgnoredDuringExecution {
            total += CountWeightByPodMatchAffinityTerm(pod, allNodes, weightedTerm.Weight, weightedTerm.PodAffinityTerm, &node)
         }
      }
      // count weights for the weighted pod anti-affinity
      if affinity.PodAntiAffinity != nil {
         for _, weightedTerm := range affinity.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution {
             total += CountWeightByPodMatchAffinityTerm(pod, allNodes, 0-weightedTerm.Weight, weightedTerm.PodAffinityTerm, &node)
         }
      }
      // "reverse" direction checking
      for _, ep := range allPods {
           epAffinity, err := api.GetAffinityFromPodAnnotations(ep.Annotations)
           // count weight for the weighted pod affinity indicated by the existing pod.
           if epAffinity.PodAffinity != nil {
                for _, weightedTerm := range epAffinity.PodAffinity.PreferredDuringSchedulingIgnoredDuringExecution {          
                   if XXX() { total += weightedTerm.weight }
                }
           }
           // count weight for the weighted pod anti-affinity indicated by the existing pod.
           if epAffinity.PodAntiAffinity != nil {
                for _, weightedTerm := range epAffinity.PodAntiAffinity.PreferredDuringSchedulingIgnoredDuringExecution {      
                   if XXX() { total += weightedTerm.weight }
                }
           }
           // count the implicit weight for the hard pod affinity indicated by the existing pod.
           // <You should be able to figure out how this works, based on the above. So I'm omitting it.>
      }
   }
   // lots of other stuff ...
}
```

`XXX()` is a function that returns true if weightedTerm.LabelSelector/Namespaces matches the labels on "pod" (the first argument to CalculateInterPodAffinityPriority())
(Note that I didn't take positive/negative into account properly to reflect affinity/anti-affinity, but I'm sure you can fix that.)